packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts

```diff
@@ -222,54 +215,110 @@ describe('testing schema definition mode `file`', () => {
   });
 
   test('definition mode `file` errors when addSubscription is called', () => {
+    // THEN
+    expect(() => {
+      api.addSubscription('blah', new appsync.ResolvableField({ returnType: t.string }));
+    }).toThrowError('Unable to add subscription. Schema definition mode must be CODE. Received: FILE');
+  });
+});
+
+describe('testing schema definition mode `s3`', () => {
+
+  test('definition mode `s3` produces correct output', () => {
+    // GIVEN
+    const bucket = new s3.Bucket(stack, 'Bucket');
+
+    new s3deploy.BucketDeployment(stack, 'DeployFiles', {
+      sources: [s3deploy.Source.asset(join(__dirname, 'schemas'))],
+      destinationBucket: bucket,
+    });
+
     // WHEN
-    const api = new appsync.GraphqlApi(stack, 'API', {
+    new appsync.GraphqlApi(stack, 'API', {
       name: 'demo',
-      schema: appsync.Schema.fromAsset(join(__dirname, 'appsync.test.graphql')),
+      schema: appsync.Schema.fromS3Location({
+        bucket,
+        key: 'appsync.test.graphql',
+      }),
     });
 
     // THEN
-    expect(() => {
-      api.addSubscription('blah', new appsync.ResolvableField({ returnType: t.string }));
-    }).toThrowError('Unable to add subscription. Schema definition mode must be CODE. Received: FILE');
+    expect(stack).toHaveResourceLike('AWS::AppSync::GraphQLSchema', {
+      DefinitionS3Location: {
+        'Fn::Join': ['',
+          [
+            's3://',
+            { Ref: 'Bucket83908E77' },
+            '/appsync.test.graphql',
+          ]],
+      },
+    });
+  });
+
+  describe('definition mode s3 fails when invoking code-first functions', () => {
+
+    let api: appsync.GraphqlApi;
+    beforeEach(() => {
+      // GIVEN
+      const bucket = new s3.Bucket(stack, 'Bucket');
+      api = new appsync.GraphqlApi(stack, 'API', {
+        name: 'demo',
+        schema: appsync.Schema.fromS3Location({
+          bucket,
+          key: 'garbage',
+        }),
+      });
+    });
+
+    test('definition mode `s3` errors when addType for an intermediate type is called', () => {
+      // THEN
+      expect(() => {
+        api.addType(new appsync.ObjectType('blah', {
+          definition: { fail: t.id },
+        }));
+      }).toThrowError('API cannot add type because schema definition mode is not configured as CODE.');
+    });
+
+    test('definition mode `s3` errors when addToSchema is called', () => {
+      // THEN
+      expect(() => {
+        api.addToSchema('blah');
+      }).toThrowError('API cannot append to schema because schema definition mode is not configured as CODE.');
+    });
+
+    test('definition mode `s3` errors when addQuery is called', () => {
+      // THEN
+      expect(() => {
+        api.addQuery('blah', new appsync.ResolvableField({ returnType: t.string }));
+      }).toThrowError('Unable to add query. Schema definition mode must be CODE. Received: S3');
+    });
+
+    test('definition mode `s3` errors when addMutation is called', () => {
```

Comment:
The descriptions of these tests could be improved. Remember that the parent descriptions are concatenated with the children. They should form proper sentences.

Something like:

```ts
  describe('API with Schema from S3', () => {
   // ...

  test('throws an exception when addQuery() is called on it', () => {
    // ...
```